core/src/main/java/org/apache/iceberg/ManifestFilterManager.java

```diff
@@ -185,6 +200,13 @@ List<ManifestFile> filterManifests(Schema tableSchema, List<ManifestFile> manife
       return ImmutableList.of();
     }
 
+    // The current set of referenced manifests can be trusted if it is a subset of the manifests
+    // being filtered. If a single referenced manifest is not in the set of manifests being filtered
+    // this indicates that the referenced manifests are stale and cannot be trusted.
+    Set<String> manifestLocations =
```

Comment:
These manifests are the set of manifests in the parent snapshot. I think the check that the referenced manifests is a subset of the manifests in the parent is sufficient. The reason is in the case of a concurrent update which removes 1 or more of the currently referenced manifests, the parent would reflect this removal (since every retry due to a concurrent update would naturally have an updated parent and therefore an updated view of what are the manifests currently). cc @rdblue @aokolnychyi

Reply:
I'm rerunning benchmarks to make sure after all the recent refactoring I did, that somehow we're not skipping the optimization.

Reply:
I think it would help to state in this comment what is meant by "trusted". I think it means that we can filter the set of referenced manifests rather than the set of manifests that was passed into this method.

Reply:
>it means that we can filter the set of referenced manifests rather than the set of manifests that was passed into this method.

Actually, I think it's more specific than this. It's possible that the set can be trusted, there's a manifest which is not referenced set, *_and_* the manifest still needs to be filtered because it either has live entries OR it's a delete manifest with aged out entries (the min sequence number ends up being less than the min data sequence number during filtering). 


This also ties in to trying to make this logic a bit less complicated since that last condition I mentioned is what's leading to the current implementation.